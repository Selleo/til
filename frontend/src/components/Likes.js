import React, { useState, useEffect } from "react";
import { useSelector } from "react-redux";
import { handleLike, checkHasLiked } from "../utils";

const Likes = props => {
  const { likesCount, id, likes } = props.post;
  const userId = useSelector(state => {
    if (!state.currentUser) {
      return null;
    } else {
      return state.currentUser.uuid;
    }
  });
  const [likesNumber, setLikesNumber] = useState(likesCount);
  const [hasLiked, setHasLiked] = useState(null);

  useEffect(() => {
    const didLike = checkHasLiked(likes, userId);

    setHasLiked(didLike);
  }, [likes, userId]);

  const toggleLike = async () => {
    if (hasLiked) {
      await handleLike(id, "DELETE")
        .then(() => {
          setHasLiked(!hasLiked);
          setLikesNumber(likesNumber - 1);
        })
        .catch(err => console.error(err));
    } else {
      await handleLike(id, "POST")
        .then(() => {
          setHasLiked(!hasLiked);
          setLikesNumber(likesNumber + 1);
        })
        .catch(err => console.error(err));
    }
  };

  return (
    <button onClick={toggleLike}>
      Likes: {likesNumber} You {hasLiked ? "like" : "dont like"} it!
    </button>
  );
};

export default Likes;
