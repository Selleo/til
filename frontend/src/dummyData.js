export const initialState = {
  categories: [],
  currentUser: null,
  users: [],
  posts: []
};

export const user = {
  email: "l.walczak@selleo.com",
  firstName: null,
  image:
    "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
  lastName: null,
  uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
};

export const allUsers = [
  {
    email: "l.walczak@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
  },
  {
    email: "maciekPlacek@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAIdsdfjdkjdfi/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-dadlfj33343443"
  }
];

export const categories = [
  { id: 1, name: "java" },
  { id: 2, name: "javascript" }
];

export const posts = [
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adsfdf",
    categoriesIds: [2, 1],
    id: 1,
    isPublic: false,
    likes: [
      {
        post_id: 1,
        user_uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
      }
    ],
    likesCount: 1,
    title: "title "
  },
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adfefsddf",
    categoriesIds: [3],
    id: 2,
    isPublic: false,
    likes: [],
    likesCount: 0,
    title: "maciekdfd"
  }
];
