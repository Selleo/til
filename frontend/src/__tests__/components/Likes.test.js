import React from "react";
import { Provider } from "react-redux";
import { render, cleanup } from "@testing-library/react";
import configureMockStore from "redux-mock-store";
import Likes from "../../components/Likes";
import { posts, user } from "../../dummyData";

describe("<Likes/>", () => {
  let component;
  let store;
  const props = { post: posts[0] };
  const mockStore = configureMockStore([]);

  beforeEach(() => {
    store = mockStore({
      currentUser: user,
    });

    component = render(
      <Provider store={store}>
        <Likes {...props} />
      </Provider>
    );
  });

  afterEach(cleanup);

  it("renders properly", () => {
    const { asFragment } = component;

    expect(asFragment()).toMatchSnapshot();
  });
});
