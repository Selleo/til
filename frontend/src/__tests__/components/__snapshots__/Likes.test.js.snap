// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Likes/> renders properly 1`] = `
<DocumentFragment>
  <button>
    Likes: 1 You like it!
  </button>
</DocumentFragment>
`;
